scripts/usb_scan.py

```diff
@@ -0,0 +1,660 @@
+                        allow_interface = True
+                        break
+                    else:
+                        log_wrapper("deny " + i.get_node())
+                        return False
+            if not allow_interface:
+                log_wrapper("deny " + i.get_node() + ", no matching rule")
+                return False
+
+        log_wrapper("allow " + device.get_node())
+        return True
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="scanner to get USB devices info")
+    parser.add_argument("-d", "--diagnostic", dest="diagnostic",
+                        action="store_true",
+                        help="enable diagnostic mode")
+    return parser.parse_args()
+
+
+def to_pusb(device):
+    """ convert UsbDevice to pusb dict
+
+    Example pusb dict:
+    [
+      {
+        "product-desc": "",
+        "product-id": "383c",
+        "description": "Lenovo_H8RU2TGE",
+        "vendor-desc": "Lenovo",
+        "version": "2.00",
+        "vendor-id": "17ef",
+        "path": "1-2",
+        "serial": "H8RU2TGE"
+      },
+      {
+        "product-desc":
+           "TransMemory-Mini / Kingston DataTraveler 2.0 Stick (2GB)",
+        "product-id": "6544",
+        "description":"Toshiba Corp._TransMemory-Mini /
+            Kingston DataTraveler 2.0 Stick (2GB)_9875B1B3F9F4CD301917964B",
+        "vendor-desc": "Toshiba Corp.",
+        "version": "2.00",
+        "vendor-id": "0930",
+        "path": "4-1.2",
+        "serial": "9875B1B3F9F4CD301917964B"
+      }
+    ]
+    :param device:(UsbDevice) the device to convert
+    :return:(dict) the key value pairs for pusb
+    """
+    pusb = {}
+
+    pusb["path"] = device.get_node()
+    # strip the leading space of "version"
+    pusb["version"] = device[UsbDevice._VERSION].strip()
+    pusb["vendor-id"] = device[UsbDevice._ID_VENDOR]
+    pusb["product-id"] = device[UsbDevice._ID_PRODUCT]
+    pusb["vendor-desc"] = device[UsbDevice._DESC_VENDOR]
+    pusb["product-desc"] = device[UsbDevice._DESC_PRODUCT]
+    # strip serial due to CA-268761
+    pusb["serial"] = device[UsbDevice._SERIAL].strip()
+
+    pusb["description"] = pusb["vendor-desc"]
+    if pusb["product-desc"]:
+        pusb["description"] += "_" + pusb["product-desc"]
+    if pusb["serial"]:
+        pusb["description"] += "_" + pusb["serial"]
+
+    return pusb
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    if args.diagnostic:
+        __DIAGNOSTIC = True
+
+    # get usb info
+    devices, interfaces = get_usb_info()
+
+    # debug info
+    log_list_wrapper(devices)
+    log_list_wrapper(interfaces)
+
+    # match interface to device
+    for i in interfaces:
+        for d in devices:
+            if i.is_child_of(d):
+                d.add_interface(i)
+
+    log_list_wrapper(devices)
+
+    # do policy check
+    policy = Policy()
+    pusbs = []
+    for d in devices:
+        if d.is_ready():
```

Comment:
why not simply

```python
if d.is_ready() and policy.check(d):
    pusbs.append(to_pusb(d))
```

We don't care of updating all the records anyway, they are thrown away after this stage

Reply:
It's for debug purpose, we can clear see if the device is allowed when print it.

Reply:
Then do
```python
if d.is_ready() and policy.check(d):
    d.set_allow(policy.check(d))
    pusbs.append(to_pusb(d))
```

Reply:
call policy.check() twice?

Reply:
Actually I think this all is overkill. We are returning only the allowed devices anyways, so let's remove the whole `.set_allow` logic from the code (the method and the attribute) and instead do:

```python
pusbs = [to_pusb(d) for d in devices if d.is_ready() and policy.check(d)]
print(json.dumps(pusbs))
```

You can add
```python
if args.diagnostic:
   disallowed = [str(to_pusb(d)) for d in devices if d.is_ready() and not policy.check(d)]
   not_ready =  [str(to_pusb(d)) for d in devices if not d.is_ready()]
   log_wrapper("Disallowed devices: {}\nNot ready devices: {}".format(
       "; ".join(disallowed),
       "; ".join(not_ready))
```
if you need to log the other things.

Reply:
done.